WS integration test infrastructure

velzevur opened the pull request:
[Pivotal story](https://www.pivotaltracker.com/story/show/154081975)
### Description
Added an utils file (`apps/aehttp/test/aehttp_ws_test_utils.erl`) for handling all the client's websocket logic. This includes some wrappers around the WS client - since by nature WS are async and this does not fit really well in tests - the functions that needed to be synchronous are sync, and the others are left async. Currently it has the support for:

* start a websocket client (sync)

* stop a websocket client (async)

* send message (async)

* subscribe/unsubscribe the test process for an event (described bellow) (sync)

* wait for a certain event (sync)

### Arguments
The main concern had been not letting race conditions inside tests. The work flow is the following:
1. Subscribe the test for certain events
2. Do something
3. Wait for the expected event

Since 1 and 3 are sync calls to the WS client process, this guarantees us that no events are lost due to races.

### Example
from ws_block_mined test
```
    ok = ?WS:subscribe_for_event(ConnPid, miner, mined_block),
    aecore_suite_utils:mine_blocks(aecore_suite_utils:node_name(?NODE), 1),
    {ok, #{<<"height">> := _Height, <<"hash">> := _Hash}} = ?WS:wait_for_event(miner, mined_block),
    ok = ?WS:unsubscribe_for_event(ConnPid, miner, mined_block),
```
We subscribe the test process to event with origin `miner` and action `mined_block`, then we mine a block and wait for the message from the WS process to arrive (default timeout is 5s).

### TODO
~~Move the starting and stopping of a WS client (being boiler plate) to the `init_per_testcase/2` and `init_per_testcase/2` respectively~~

- [x] A WS connection can subscribe to certain events (for example _this_ oracle events). This functionality is not implemented yet in the test client. This would also require some renaming - we need to distinguish test subscribes to expected events with WS connection's subscriptions. Probably test's subscribtions would be renamed to registering or something like that. Comments are welcome.
- [x] Write some WS tests to cover the current functionality

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -1309,6 +1323,34 @@ block_txs_list_by_hash_invalid_range(_Config) ->
+ws_get_genesis(_Config) ->
+    {Host, Port} = ws_host_and_port(),
+    {ok, ConnPid} = ?WS:start(Host, Port),
+    ?WS:send(ConnPid, chain, get, #{height => 0}),
```
Could you please ensure this is [documented](https://github.com/aeternity/protocol/blob/68b27350f157e05c6930d037a626695e6a4fad3b/epoch/api/README.md#websocket-api-definition) ?
(It may require moving out of the oracle page some more generic portions.)

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -1309,6 +1323,34 @@ block_txs_list_by_hash_invalid_range(_Config) ->
+ws_block_mined(_Config) ->
+    {Host, Port} = ws_host_and_port(),
+    {ok, ConnPid} = ?WS:start(Host, Port),
+    ok = ?WS:subscribe_for_event(ConnPid, miner, mined_block),
```
Could you please ensure this is [documented](https://github.com/aeternity/protocol/blob/68b27350f157e05c6930d037a626695e6a4fad3b/epoch/api/README.md#websocket-api-definition) ?
(It may require moving out of the oracle page some more generic portions.)

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    ssl:start(),
+    WsAddress = "ws://" ++ Host ++ ":" ++ integer_to_list(Port) ++ "/websocket",
+    ct:log("connecting to ~p", [WsAddress]),
+    {ok, Pid} = websocket_client:start_link(WsAddress, ?MODULE, self()),
```
This is actually linking. Could you please consider renaming from start to start_link ?

lucafavatella replied:
It does not matter much, though websocket_client:start_link types 3rd arg as list.

uwiger commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+
+stop(ConnPid) ->
+    ct:log("Websocket disconnecting"),
+    ConnPid ! stop.
```
Perhaps you could monitor `ConnPid` and wait for a `'DOWN'` message?

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+
+start(Host, Port) ->
+    crypto:start(),
+    ssl:start(),
```
I see these are application:start of crypto and ssl. Shall these two lines be removed from here and moved to aehttp_integration_SUITE:init_per_group websocket? (And maybe corresponding stops in end_per_group.)
Please consider (I am not sure what current design in tests is):
* Matching return value (so to fail at correct point if e.g. crypto cannot start);
* Using start/ensure_started/ensure_all_started and stop.

BTW I see these crypto an ssl are [dependencies of websocket_client](https://github.com/sanmiguel/websocket_client/blob/2b6228211389fbd0b1694e5b71e33da48d98d2b7/src/websocket_client.app.src#L5)

lucafavatella replied:
I see - originally taken from https://github.com/sanmiguel/websocket_client/tree/2b6228211389fbd0b1694e5b71e33da48d98d2b7#usage

lucafavatella commented on apps/aehttp/src/ws_int_dispatch.erl:
```diff
@@ -28,7 +28,7 @@ is_valid(_, _, _) ->
 
 do_execute(chain, get, QueryPayload) ->
-    #{<<"type">> := Type} = QueryPayload,
+    Type = maps:get(<<"type">>, QueryPayload, <<"block">>),
```
Why a default? I would rather expect to keep this code simpler not having the default, documenting what is needed (in protocol repo) and requiring client to comply to documented schema.
(I may not understand the problem here so please correct me if wrong.)

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+
+-spec wait_for_event(atom(), atom()) -> ok | {ok, map()} | timeout.
+wait_for_event(Origin, Action) ->
+    wait_for_event(Origin, Action, 5000).
```
Macros for timeouts?

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+unsubscribe_for_event(ConnPid, Origin, Action) ->
+    Event = {Origin, Action},
+    ConnPid ! {unsubscribe, self(), Event},
+    ok = wait_for_event(unsubscribed, Event, 1000),
```
Macros for timeouts?

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    WsAddress = "ws://" ++ Host ++ ":" ++ integer_to_list(Port) ++ "/websocket",
+    ct:log("connecting to ~p", [WsAddress]),
+    {ok, Pid} = websocket_client:start_link(WsAddress, ?MODULE, self()),
+    ok = wait_for_event(websocket, connected),
```
I see - on init argument (self()) is automatically subscribed to waiting_connected, and onconnect subscription is removed and message is sent to caller. This may lead to unexpected message in mailbox of caller of aehttp_ws_test_utils:start (in case of timeout/slowness/race). It would be nice to have a line of comment documenting this potential message in.

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    Payload = maps:get(<<"payload">>, Msg, []),
+    Subscribed = get_subscribed({Origin, Action}, Subs),
+    lists:foreach(
+        fun(Pid) -> inform_subscribed(Pid, Origin, Action, Payload) end,
```
I see this is a real asynchronous notification.

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    websocket_client:cast(ConnPid, {text, jsx:encode(Msg)}),
+    ok.
+
+subscribe_for_event(ConnPid, Origin, Action) ->
```
Could you please add a line of comment mentioning that caller will receive in mailbox messages looking like {websocket_event, ...} (or similar - better if with a type).

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    {close, <<>>, "stopped"};
+websocket_info({subscribe, SubscribedPid, Event}, _ConnState, Subs0) ->
+    Subs = put_subscription(SubscribedPid, Event, Subs0),
+    inform_subscribed(SubscribedPid, subscribed, Event),
```
> Would it be possible to use websocket_client:send

It looks no, as it frames the message in `{send, _}`

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,155 @@
+    {ok, Subs};
+websocket_info({unsubscribe, SubscribedPid, Event}, _ConnState, Subs0) ->
+    Subs = delete_subscription(SubscribedPid, Event, Subs0),
+    inform_subscribed(SubscribedPid, unsubscribed, Event),
```
> Would it be possible to use websocket_client:send

It looks no, as it frames the message in `{send, _}`

velzevur commented:
I think I've addressed all comments. PR for protocol update to current state of WS chain interaction can be found [here](https://github.com/aeternity/protocol/pull/21)
I'm proceeding with renaming of the test's subscription to certain events to `register`-ing

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -147,7 +156,6 @@ groups() ->
       ]},
      {internal_endpoints, [sequence], 
       [
-        broken_spend_tx,
```
Spurious change?

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -209,36 +223,21 @@ end_per_suite(_Config) ->
+init_per_group(_Group, Config) ->
     aecore_suite_utils:start_node(?NODE, Config),
     aecore_suite_utils:connect(aecore_suite_utils:node_name(?NODE)),
-    [{start_node_per_tc, false} | Config];
```
I see this is a test simplification. Ok.

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -209,36 +224,21 @@ end_per_suite(_Config) ->
     Config;
-init_per_group(Group, Config) when Group =:= external_endpoints
-                            orelse Group =:= internal_endpoints ->
+init_per_group(_Group, Config) ->
```
I expect that this change, because of the group-nesting group `all_endpoints`, makes this function body to be run twice (without the corresponding `end_per_group` having been run already). So the requirement on this function body is that it is idempotent. It is idempotent at the moment, as the node start can be run again and it leaves the node running, and and the aehttp app_started event is recorded in the proxy. Though for avoiding future breakages I suggest to either:
A. Restore start of node only for specified leaf groups (external_endpoints, internal_endpoints, websockets); or
B. Add comment that body here needs to be idempotent.

I recommend A.

lucafavatella replied:
> A. Restore start of node only for specified leaf groups (external_endpoints, internal_endpoints, websockets); or

I had not noticed that clause before already handles `all_endpoints`, so please ignore me on this. Thanks for pointing that out.

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,165 @@
+%% {websocket_event, Origin::atom(), Action::atom()}
+%% {websocket_event, Origin::atom(), Action::atom(), Payload::map()}
+%% where the Origin and Action are the same as the ones for which the process
+%% had registered. This message is consumed by wait_for_event/2,3
```
This means a test process cannot start two WS clients and discriminate the received messages by source WS connection. Though not needed for current tests so ok.

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -1309,6 +1311,68 @@ block_txs_list_by_hash_invalid_range(_Config) ->
+            end,
+            lists:seq(1, AllowedWsCount)),
+    AllowedWsCount = open_websockets_count(), 
+    {error, timeout} = ws_start_link(),
```
I understand the exercised timeout is DEFAULT_SUB_TIMEOUT i.e. 1000 (1s). Ok.

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -0,0 +1,169 @@
+        ok ->
+            {ok, Pid};
+        timeout ->
+            {error, timeout}
```
When this happen, does WS client slow to connect (potentially for server that ran out of acceptors) stay lingering expecting to connect? Hence eventually consuming connection? Hence eventually potentially putting unexpected message in mailbox of test process?

velzevur commented:
Enchanced the test for verifying the maximum amount of connected WS clients (and that we actually queue the next X incoming ones)

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -1353,24 +1353,65 @@ ws_block_mined(_Config) ->
+            %% stop one
+            ?WS:stop(Pid),
+            %% another one connects in its place and it doesn't matter which one
+            {ok, some_pid} = ?WS:wait_for_connect(some_pid),
```
Can this `some_pid` work? Maybe SomePid?

lucafavatella replied:
oh not used - I see...

lucafavatella commented on apps/aehttp/test/aehttp_integration_SUITE.erl:
```diff
@@ -1353,24 +1353,65 @@ ws_block_mined(_Config) ->
+        FirstWSsPids),
+    %% cleanup
+    lists:foreach(fun ?WS:stop/1, OtherWSsPids),
+    WSQueueSize = open_websockets_count(), 
```
Started MaxWsCount+WSQueueSize. Then stopped WSQueueSize. I would expect MaxWsCount left?

lucafavatella replied:
[CI failed](https://travis-ci.org/aeternity/epoch/jobs/333152759), there may be some miscalculation here indeed (actually test failed 2 lines below...).

velzevur replied:
Indeed started MaxWsCount+WSQueueSize
Then stopped `WSQueueSize` and `OtherWSsPids` (where OtherWSsPids = MaxWsCount - WSQueueSize, the WS-s that were not stopped while stopping the first WSQueueSize number of WS-s). Since WSQueueSize + OtherWSsPids = MaxWsCount, we have actually stopped MaxWsCount amount of WS-s, leaving us with just WSQueueSize left. I think the test assumption is correct

lucafavatella commented on apps/aehttp/test/aehttp_ws_test_utils.erl:
```diff
@@ -109,9 +115,12 @@ onconnect(_WSReq, Regs) ->
     {ok, Regs1}.
 
+
+ondisconnect({error, {400, <<"Bad Request">>}}, Regs) ->
```
When does this happen? Worth documenting in protocol repo?

velzevur replied:
When all Websocket acceptors are consumed and the queue for incoming new Websockets is full, we start rejecting new connections. This is the standart `cowboy` message being send to the request. It is not documented in the protocol repo, I'm not sure if it would add more clarity or the opposite.

lucafavatella replied:
mmm it could surprise some clients so I would consider it worth documenting it
